improver/expected_value.py

```diff
@@ -30,21 +30,108 @@
 # POSSIBILITY OF SUCH DAMAGE.
 """Calculation of expected value from a probability distribution."""
 
+import numpy as np
 from iris.coords import CellMethod
 from iris.cube import Cube
 
 from improver import PostProcessingPlugin
 from improver.ensemble_copula_coupling.ensemble_copula_coupling import (
-    ConvertProbabilitiesToPercentiles,
     RebadgePercentilesAsRealizations,
+    get_bounds_of_distribution,
+)
+from improver.metadata.probabilistic import (
+    find_threshold_coordinate,
+    is_percentile,
+    is_probability,
 )
-from improver.metadata.probabilistic import is_percentile, is_probability
 from improver.utilities.cube_manipulation import collapse_realizations
+from improver.utilities.probability_manipulation import to_threshold_inequality
 
 
 class ExpectedValue(PostProcessingPlugin):
     """Calculation of expected value from a probability distribution"""
 
+    @staticmethod
+    def integrate_over_thresholds(cube: Cube) -> Cube:
+        """
+        Calculation of expected value for threshold data by converting from
+        cumulative distribution (CDF) to probability density (PDF), then
+        integrating over the threshold dimension.
+
+        Args:
+            cube:
+                Probabilistic data with a threshold coordinate.
+
+        Returns:
+            Expected value of probability distribution. Same shape as input cube
+            but with threshold coordinate removed.
+        """
+        # make sure that the threshold direction is "below"
+        cube = to_threshold_inequality(cube, above=False)
+        # set up threshold values, these will be needed as a multiplier during integration
+        threshold_coord = find_threshold_coordinate(cube)
+        threshold_coord_idx = cube.dim_coords.index(threshold_coord)
+        thresholds = threshold_coord.points
+        # add an extra threshold below/above with zero/one probability
+        # this ensures the PDF integral covers the full CDF probability range
+        try:
+            # use bounds from ECC as an outer bound on the distribution if available
+            ecc_bounds = get_bounds_of_distribution(
+                threshold_coord.name(), threshold_coord.units
+            )
+        except KeyError:
+            # use infinities to fall back to threshold_spacing if ECC bounds missing
+            ecc_bounds = np.array([np.inf, -np.inf])
```

Comment:
This is intentional.
- If there are ECC bounds based on the name lookup, the min/max of those bounds and the threshold spacing is used. This results in the newly added thresholds always covering the ECC bounds range and always extending the existing threshold range
- If no ECC bounds are found, the threshold spacing method is used.

Thresholds of +/- infinity would cause problems for numerical integration, such as calculating the midpoints a few lines further down.

Reply:
Good point, I hadn't considered what would happen when you try and take a mid-point when one of the values is +/-np.inf (this is the point where the numerical integration breaks down as you mention).

I'm happy to resolve this, but one suggestion might be to add a comment here to avoid the help the reader avoid my initial (naive) interpretation (I'll leave this one up to you).

Reply:
Thanks for adding further clarification here.